src/Draco.Lsp/Server/LanguageServer.cs

```diff
@@ -45,62 +37,42 @@ public static ILanguageClient Connect(Stream stream)
     /// <returns>The task that completes when the communication is over.</returns>
     public static async Task RunAsync(this ILanguageClient client, ILanguageServer server)
     {
-        var jsonRpc = client.Connection;
+        var connection = client.Connection;
 
-        // Register builtin server methods
-        RegisterBuiltinRpcMethods(server, jsonRpc);
         // Register server methods
-        RegisterServerRpcMethods(server, jsonRpc);
+        RegisterServerRpcMethods(server, connection);
 
-        // Done, now we can actually start
-        jsonRpc.StartListening();
-        await jsonRpc.Completion;
-    }
+        if (server is not ILanguageServerLifecycle)
+        {
+            // Register builtin server methods
+            var lifecycle = new LanguageServerLifecycle(server, connection);
+            RegisterServerRpcMethods(lifecycle, connection);
+        }
```

Comment:
Ah this might explain my confusion, we optionally allow overriding lifecycle methods?

Reply:
The interface should probably be internal. I added it because `RegisterServerRpcMethods` will only add methods defined on interfaces. The interface itself is not useful for extensibility, since `LanguageServerLifecycle` implements capability registration. So a consumer who wants to provide their own `ILanguageServerLifecycle` would need to reimplement that from scratch. If we wanted to do something here it'd require more thought